cni/pkg/taint/taintcontroller_test.go

```diff
@@ -0,0 +1,498 @@
+		}
+		tempMap, _ := metav1.LabelSelectorAsMap(labelselector)
+		for key, val := range tempMap {
+			labelMap[key] = val
+		}
+	}
+	var condition v1.PodCondition
+	if podArg.readiness {
+		condition = v1.PodCondition{
+			Type:   v1.PodReady,
+			Status: v1.ConditionTrue,
+		}
+	} else {
+		condition = v1.PodCondition{
+			Type:   v1.PodReady,
+			Status: v1.ConditionFalse,
+		}
+	}
+	return makePodWithTolerance(makePodArgs{
+		PodName:             podArg.podName,
+		Namespace:           podArg.namespace,
+		Labels:              labelMap,
+		InitContainerStatus: &workingInitContainer,
+		Tolerations: []v1.Toleration{
+			{Key: TaintName, Operator: v1.TolerationOpExists, Effect: v1.TaintEffectNoSchedule},
+		},
+		NodeName: podArg.nodeName,
+		Conditions: []v1.PodCondition{
+			condition,
+		},
+	})
+}
+
+type nodeInfo struct {
+	nodeName  string
+	hasTaint  bool
+	readiness bool
+}
+
+func mockNodeGenerator(nodeArgs nodeInfo) *v1.Node {
+	var taint v1.Taint
+	if nodeArgs.hasTaint {
+		taint = v1.Taint{Key: TaintName, Effect: v1.TaintEffectNoSchedule}
+	} else {
+		taint = v1.Taint{}
+	}
+	var nodeReadiness v1.NodeCondition
+	if nodeArgs.readiness {
+		nodeReadiness = v1.NodeCondition{
+			Type:              v1.NodeReady,
+			Status:            v1.ConditionTrue,
+			LastHeartbeatTime: metav1.Time{Time: time.Unix(1, 1)},
+		}
+	} else {
+		nodeReadiness = v1.NodeCondition{
+			Type:              v1.NodeReady,
+			Status:            v1.ConditionFalse,
+			LastHeartbeatTime: metav1.Time{Time: time.Unix(1, 1)},
+		}
+	}
+	return makeNodeWithTaint(makeNodeArgs{NodeName: nodeArgs.nodeName, Taints: []v1.Taint{taint}, NodeCondition: []v1.NodeCondition{nodeReadiness}})
+}
+func TestController_ListAllNode(t *testing.T) {
+	type fields struct {
+		client kubernetes.Interface
+	}
+	type args struct {
+		nodeNames []string
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+		want   map[string]bool
+	}{
+		{
+			name: "add a node",
+			fields: fields{
+				client: fakeClientset([]v1.Pod{}, []v1.Node{testingNode}, []v1.ConfigMap{})},
+			args: args{nodeNames: []string{"a"}},
+			want: map[string]bool{"a": true},
+		},
+		{
+			name: "add several node",
+			fields: fields{
+				client: fakeClientset([]v1.Pod{}, []v1.Node{testingNode}, []v1.ConfigMap{})},
+			args: args{nodeNames: []string{"a", "b", "c"}},
+			want: map[string]bool{"a": true, "b": true, "c": true},
+		},
+	}
+	for _, tt := range tests {
+		ts := TaintSetter{configs: []ConfigSettings{}, Client: tt.fields.client}
+		source := fcache.NewFakeControllerSource()
+		tc, _, err := newMockTaintSetterController(&ts, source)
+		if err != nil {
+			t.Fatalf("cannot constrict taint controller")
+		}
+		stop := make(chan struct{})
+		go tc.Run(stop)
+		for _, name := range tt.args.nodeNames {
+			source.Add(&v1.Node{ObjectMeta: metav1.ObjectMeta{Name: name}})
+		}
+		// Let's wait for the controller to finish processing the things we just added.
+		time.Sleep(100 * time.Millisecond)
```

Comment:
do not sleep, use `retry.UntilSucess` instead to poll. Sleeping is not safe, CI runs in a shared environment and gets throttled which may make sleep not long enough. If you make it longer, the test is too slow

Reply:
removed all sleeps